Add Cupertino Countdown Timer Picker

doxuanviet1996 opened the pull request.

amirh committed:
Integration test for embeded Android Views touch support.

The test places an embedded Android view at the top left, and verifies
that motion events that get to FlutterView are equivalent to the
synthesized motion events that gets to the embedded view.

See the README.md for more high level details.

amirh committed:
flutter update-packages

amirh committed:
Clone goldens as part of flutter update-packages.

flutter_tools cannot depend on flutter_goldens (as flutter_goldens
depdends on the Flutter sdk), so this commit splits client.dart from
flutter_goldens to a pure-dart flutter_goldens_client package.

doxuanviet1996 committed:
add countdown timer class

doxuanviet1996 committed:
add countdown timer class

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

DaveShuckerow commented on examples/flutter_gallery/lib/demo/cupertino/cupertino_picker_demo.dart:
```diff
@@ -20,8 +20,7 @@ class CupertinoPickerDemo extends StatefulWidget {
 
-  int _selectedHour = 0;
-  int _selectedMinute = 0;
+  Duration timer = new Duration(minutes: 1);
```
You'll need to explain what this is for if you decide to submit this demo.

DaveShuckerow commented on packages/flutter/lib/src/cupertino/date_picker.dart:
```diff
@@ -0,0 +1,185 @@
+
+/// A countdown timer picker in iOS style.
+///
+/// This picker shows duration as hour and minute spinners. The minimum duration
```
This section is very clear!

DaveShuckerow commented on packages/flutter/lib/src/cupertino/date_picker.dart:
```diff
@@ -0,0 +1,185 @@
+  Widget _buildMinutePicker(BuildContext context) {
+    return new CupertinoPicker(
+      scrollController: minuteController,
+      offAxisFraction: 0.5,
```
Note: when you try out different text scales, it's possible that will affect how this should look. If that's the case, you may need to write a method to calculate what these values should be.

DaveShuckerow commented on packages/flutter/lib/src/cupertino/date_picker.dart:
```diff
@@ -0,0 +1,185 @@
+        children: <Widget>[
+          new Row(
+            children: <Widget>[
+              new Expanded(child: _buildHourPicker(context)),
```
You'll need to figure out if this order might change in different locales. If it does, this layout will be difficult to reconfigure.

DaveShuckerow commented on packages/flutter/lib/src/cupertino/date_picker.dart:
```diff
@@ -0,0 +1,185 @@
+                child: new Container(
+                  alignment: const Alignment(0.6, 0.0),
+                  child: const Text(
+                    'hours',
```
I'm concerned about placing the labels in an overlay on top of the pickers, because changes in the size of the labels should affect the underlying pickers too.

Making sure they're consistent with each other could be complicated.

DaveShuckerow reviewed (commented):
So far so good. Here are my initial thoughts.

Hixie commented on packages/flutter/lib/src/cupertino/date_picker.dart:
```diff
@@ -0,0 +1,185 @@
+      ),
+    );
+  }
+}
```
this needs to be localized, support rtl, and support accessibility

Hixie replied:
also it will need tests

Commit:
Add  semantics to cupertino slider (#20476)

gspencergoog committed:
Fix access of null compiler in flutter tests that fail before creating the compiler. (#20553)

In certain cases, the test would fail before creating the (lazily created) compiler object, and then we'd
try to call shutdown() on null in those cases.

Fixes #18610

Commit:
add label to drawer and regression test cases (#20537)

leoylung committed:
Check null on componentName (#20560)

gspencergoog committed:
Pushing empty commit to kick all the devicelab bots

gspencergoog committed:
Pushing empty commit to kick all the devicelab bots (again)

tvolkert committed:
Pass icu data file to fuchsia_tester (#20605)

Commit:
Revert to defaultClipBehavior (#20578)

Hixie committed:
Shim package:test to avoid matcher issues (#20602)

* Upgrade everything except matcher.
* Roll matcher (and test)
* Adjust tests that depend on flutter:test directly to depend on a shim
* Require use of package:test shim and remove other references to package:test

xster committed:
Follow up on #18424 comments (#20125)

cmyr committed:
Fix link to input_event_constants.fidl (#20588)

nsampsell committed:
Add CupertinoActionSheet to gallery (#20411)

mkustermann committed:
Roll engine to version b148e628ec86b3a9a0382e0bcfae73f0390a8232 (#20427)

This CL

  * rolls `engine.version` to flutter/engine@b148e628 (which includes dart sdk 2.1.0-dev)
  * rolls `goldens.version` to flutter/goldens@6c45fafdf (which includes updates due to skia changes in engine)
  * changes `platform.dill` to `platform_strong.dill` in various places due to flutter/engine@a84b210b
  * adds explicit `environment: sdk: ">=2.0.0-dev.68 < 3.0.0"` constraints to `pubspec.yaml` and `pubspec.yaml.tmpl` files (since pub defaults to `<2.0.0` if omitted) 
  * upgrades to newer versions of various 3rd party packages (to ensure transitive dependencies have `<3.0.0` sdk constraint)

mkustermann committed:
Revert "Roll engine to version b148e628ec86b3a9a0382e0bcfae73f0390a8232 (#20427)" (#20631)

Reason for revert: The package:flutter_gallery_assets has removed some images which are required for the examples/flutter_gallery, so the gallery build is failing (only discovered after landing, since gallery doesn't seem to get built during github PR presubmit checks)

awdavies committed:
[frdb] Adds function to await Dart VM's (#20597)

In the event that no Dart VM's exist or are connectable, will instead
await for a new Dart VM to show up. Will no longer return null when
calling `getMainIsolatesByPattern`, instead returning either a list of
isolates or timing out.

goderbauer committed:
Always honor explicitly set keyboardType and inputAction for TextField (#20547)

With this change you can now create "wrapping" text fields that do not show the new line button:

```dart
new TextField(
  maxLines: null,
  keyboardType: TextInputType.text,
);
```

Commit:
Add tests/matchers for automatic accessibility testing (#20462)

\

Hixie committed:
Implement `flutter test -j` (#20493)

sbaranov committed:
Flutter tool support for dynamic code. (#20543)

matthew-carroll committed:
iOS Dialog blur, brightness, and layout (#18381)

Rewrote CupertinoAlertDialog to look nearly identical to an alert dialog in iOS. This includes considerations for blur, translucent white background color, button sizing, gap dividers between buttons, and text scaling layout behavior. (#18381)

Commit:
Revert "Flutter tool support for dynamic code" (#20644)

* Revert "iOS Dialog blur, brightness, and layout (#18381)"

This reverts commit 21bc9f1b02f657d156980adf2b7436e9872cae35.

* Revert "Flutter tool support for dynamic code. (#20543)"

This reverts commit 25ba90aafadce24dd97c55442d011b9352b1e468.

doxuanviet1996 committed:
accessibility fix

doxuanviet1996 committed:
accessibility fix

jonahwilliams commented:
@doxuanviet1996  See https://github.com/flutter/flutter/pull/20653 for pending work to match iOS accessibility behavior. Not done yet, but worth checking out to make sure this implementation behaves.

mkustermann committed:
Reland "Roll engine to version e3687f70c7ece72000b32ee1b3c02755ba5361ac (#20427)" (#20633)

* Reland "Roll engine to version b148e628ec86b3a9a0382e0bcfae73f0390a8232 (#20427)"

This is a re-land with downgraded `package:flutter_gallery_assets`
version.

* Downgrade package:flutter_gallery_assets to 0.1.4

* Change engine.version to 81baff97c29bb08cbf8453a3f9042c5813f84ad3 (which contains an additional fix)

* Change engine.version to e3687f70c7ece72000b32ee1b3c02755ba5361ac (since mac tarballs are corrupted on earlier commit)

goderbauer committed:
Copy flutter.bat before execution (#20554)

xster committed:
Correctly chain finders to ancestor/descendant finders (#20601)

xster committed:
Add doc describing sliver children's lifecycle and persistence options (#19795)

chinmaygarde committed:
Fix type mismatch while taking a screenshot using the resident runner. (#20555)

xster committed:
Create a ValueListenableBuilder (#19729)

goderbauer committed:
Revert "Copy flutter.bat before execution" (#20678)

mravn-google committed:
Materialize Flutter module, Android (#20520)

mravn-google committed:
Avoid having flutter_tools test outcome depend on packages get (#20683)

mravn-google committed:
Fix compiler error (#20687)

engine-flutter-autoroll committed:
Roll engine e3687f70c7ec..4ee648914f91 (9 commits) (#20696)

Auto-roller completed checks. Merging.

sbaranov committed:
Reland Flutter tool support for dynamic code #20543 (#20646)

sbaranov committed:
Replace flutter --build-snapshot with --precompile that takes input (#20574)

sbaranov committed:
Remove unused snapshot.d depfile parameter (#20584)

aam committed:
Add '-t' option to 'attach' command. (#20539)

* Add '-t' option to 'attach' command.

* Add test

* Make analyzer happy

* Fix tests so they use memory file system and can find lib/main.dart

xster committed:
make overflow dots on iOS horizontal in gallery (#20702)

xster committed:
copy edit modal sheet demo (#20703)

xster committed:
use the <> icon to show code examples in the gallery (#20705)

doxuanviet1996 committed:
design fix

doxuanviet1996 committed:
design fix

doxuanviet1996 committed:
design fix

doxuanviet1996 committed:
design fix

doxuanviet1996 committed:
bug fix

doxuanviet1996 committed:
bug fix

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

doxuanviet1996 committed:
doc fixes

Commit:
Add support for by word movement granularities to semantics (#19949)

Commit:
fix dart analysis error (#20698)

Hixie committed:
Try to resolve an intermitted crash during coverage collection (#20506)

* Try to resolve an intermitted crash during coverage collection

The only theory I can come up with is that maybe the test completes
before we finish processing the standard input, so I made the test
harness wait for the observatory URL before considering whether the
test has finished or not.

Also, some code cleanup while I'm at it, e.g. avoiding using "onFoo"
for the names of methods, avoiding back-to-back switch statements with
the same values, avoiding `_` argument names, and using `?.` instead
of `if (foo != null) foo.`.

* Revert back the signature of _pipeStandardStreamsToConsole

* Also remove the other additions to this method.

Hixie committed:
Clean up usage of temporary directories (#20682)

All temporary directory start with `flutter_` and have their random component separated from the name by a period, as in `flutter_test_bundle.YFYQMY`.

I've tried to find some of the places where we didn't cleanly delete temporary directories, too. This greatly reduces, though it does not entirely eliminate, the directories we leave behind when running tests, especially `flutter_tools` tests.

While I was at it I standardized on `tempDir` as the variable name for temporary directories, since it was the most common, removing occurrences of `temp` and `tmp`, among others.

Also I factored out some common code that used to catch exceptions that happen on Windows, and made more places use that pattern.

Hixie committed:
Track number of package dependencies in Flutter (#20722)

Hixie committed:
Performance test cleanup (#20652)

* Fix TODO syntax.
* Clarify messages for some timeouts, to aid debugging.
* Increase some other timeouts that were a needlessly short, to reduce sources of flakes.
* Remove some more timeouts that were mostly redundant, to remove complexity.
* Minor style cleanup.
* Remove some dangerous traps (specifically, hide the explicit start/end times in TimedEvent since they shouldn't matter).

jcollins-g committed:
Add branch to footer information in flutter docs (#20711)

Hixie committed:
Documentation regarding tap gesture callbacks (#20647)

jcollins-g committed:
Use markdown table (#20721)

doxuanviet1996 committed:
temporary save

doxuanviet1996 committed:
temporary save

tianlunlee committed:
Feature pesto new recipes (#19415)

<img width="273" alt="pesto_before" src="https://user-images.githubusercontent.com/13839358/43009716-64b6d726-8c0c-11e8-989a-13eaff72ccee.png">
<img width="273" alt="pesto_after" src="https://user-images.githubusercontent.com/13839358/43009717-66074a34-8c0c-11e8-92ae-f5fb4074f43a.png">

doxuanviet1996 committed:
temporary save

doxuanviet1996 committed:
temporary save

matthew-carroll committed:
Deprecated CupertinoDialog in favor of a new widget called CupertinoP… (#20649)

Deprecated CupertinoDialog in favor of a new widget called CupertinoPopupSurface. (#20397)

doxuanviet1996 committed:
use localization

doxuanviet1996 committed:
use localization

doxuanviet1996 committed:
use localization

doxuanviet1996 committed:
use localization

doxuanviet1996 committed:
textDirection fix

doxuanviet1996 committed:
textDirection fix

Hixie committed:
Revert "Track number of package dependencies in Flutter" (#20774)

* Revert "Deprecated CupertinoDialog in favor of a new widget called CupertinoP… (#20649)"

This reverts commit 3a4ae280e3a4af0f93709bac7a87a1ce5445533a.

* Revert "Feature pesto new recipes (#19415)"

This reverts commit e2f3b3d6f2c164975c0ec0d81345c9f603cd2d44.

* Revert "Use markdown table (#20721)"

This reverts commit 82d43b952ac3fc8c412c850f105710e55e6df565.

* Revert "Documentation regarding tap gesture callbacks (#20647)"

This reverts commit 3acc2785219e61cba73dc5540a9a50360d7716f0.

* Revert "Add branch to footer information in flutter docs (#20711)"

This reverts commit 9118d3d715bf2f3407a1146b7c6585abe7d031a5.

* Revert "Performance test cleanup (#20652)"

This reverts commit 1993a67381f03a8d472da687cd2c1582e9f788a1.

* Revert "Track number of package dependencies in Flutter (#20722)"

This reverts commit 03d6f18f4a900e0e7f94060b54042365b00ad069.

Hixie committed:
Fixes resulting from audit of issues links (#20772)

* Fixes resulting from audit of issues links

I looked at every link to GitHub in our repo. For cases where we had a TODO that was waiting for a bug to be fixed, and the bug has now been fixed, I applied the pending change. For cases where the link was out of date, I updated the link.

* Update run_test.dart

skip this test again since it failed on linux and macos bots

Hixie committed:
Delete all temporary files in flutter test. (#20679)

Turns out we weren't deleting the dill directory or the fonts directory.

mehmetf committed:
Allow option switches to have keys in Gallery (#20740)

* Allow option switches to have keys

* Review comments

doxuanviet1996 committed:
add test

doxuanviet1996 committed:
add test

doxuanviet1996 committed:
add tests

doxuanviet1996 committed:
add tests

engine-flutter-autoroll committed:
Roll engine 4ee648914f91..03f8ba51e3d4 (13 commits) (#20790)

Auto-roller completed checks. Merging.

engine-flutter-autoroll committed:
Roll engine 03f8ba51e3d4..4e6f57949d2c (3 commits) (#20817)

Auto-roller completed checks. Merging.

amirh committed:
Canonicalize path depdendencies in `flutter update-packages`

The tool was failing if we had 2 separate non-canonical paths to the
same canonical paths.

engine-flutter-autoroll committed:
Roll engine 4e6f57949d2c..9cf80283cd73 (1 commits) (#20823)

Auto-roller completed checks. Merging.

DaveShuckerow committed:
Wrap the ReorderableList's children with a MergeSemantics (#20715)

engine-flutter-autoroll committed:
Roll engine 9cf80283cd73..c23690f0d751 (1 commits) (#20825)

Auto-roller completed checks. Merging.

Commit:
Fix (another) typo in InputDecorationTheme.debugFillProperties() (#20759)

doxuanviet1996 committed:
timer picker fix

doxuanviet1996 committed:
timer picker fix

Hixie committed:
`flutter analyze` cleanup (#20490)

* `flutter analyze` cleanup

* Make `--dartdocs` work in all modes.
* Make `analyze-sample-code.dart` more resilient.
* Add a test for `analyze-sample-code.dart`.
* Minor cleanup in related code and files.

* Apply review comments

* Fix tests

liyuqian committed:
Add golden test for clipping (#20750)

For #20483

zanderso committed:
Roll engine to e26643faf83f865deda65a68b2342bbd46c96093

Hixie committed:
Add ignore_for_file to the technical debt menu (#20755)

If //ignore is bad, ignoring something for a whole file is presumably worse.

engine-flutter-autoroll committed:
Roll engine e26643faf83f..4e845be0b1ac (5 commits) (#20848)

Auto-roller completed checks. Merging.

lfkdsk committed:
release lock (#19671)

engine-flutter-autoroll committed:
Roll engine 4e845be0b1ac..4b271b2e0265 (3 commits) (#20873)

Auto-roller completed checks. Merging.

ebeem committed:
Add ability to set collapse mode in flexible space bar (#18526)

* Add ability to set collapse mode in flexible space bar

* Add tests to collapse mode in flexible space bar

* Fix minor style nit

* Fix getting collapse mode does not handle unknown values

* Change const key to final in space bar collapse mode test

* Return null collapse padding if mode is unknown

devoncarew committed:
add a --use-cfe option to flutter analyze (#20742)

* add a --use-cfe option to flutter analyze

* useCFE ==> useCfe

DanTup committed:
Provide better error messages in integration tests when apps don't launch successfully (#20813)

* Don't keep parsing results once we've had the event we expected

* Report errors if an app.stop event is received while waiting for another event

* Don't throw if we see app.stop event when we've just sent an app.stop request

* Improve debug print to include --start-paused if being used

* Improve wrapping

liyuqian committed:
Clip the flutter_gallery front layer (#20836)

This will be needed once PhysicalShape sets the default clipBehavior
to Clip.none.

leoylung committed:
Remove duplicated assets under /pkg/data (#20728)

Require componentName to ensure assets are placed under /pkg/data/<component-name>

Hixie committed:
Audit `TODO` syntax (#20837)

Fixes the pattern for some TODOs to match our style guide.

(Also, a couple of minor code order fixes.)

Hixie committed:
Track number of package dependencies in Flutter (#20831)

* Track number of package dependencies in Flutter

Relands #20774.

* Use evalFlutter instead of startProcess.

That way we don't need `flutter` on the PATH.

Hixie committed:
Cleanup (#20754)

- remove an //ignore that is no longer needed
- fix some intrinsic methods that used the API incorrectly (shouldn't affect correctness but should make things a tiny bit more efficient)
- add some asserts to help track down bugs quicker
- update a TODO to point to the currently relevant bug
- fix some indenting
- improve the naming of some privates to improve readability

Commit:
Update BottomNavigationBar.didUpdateWidget() (#20890)

BottomNavigationBar's state needs to update _backgroundColor when its configuration changes.

DanTup committed:
Double all timeouts on flutter_tools integration tests (#20872)

Due to CPU contention we've seen these go really slow on Cirrus (see https://github.com/flutter/flutter/issues/19542#issuecomment-414265152), and there's also a chance our flakes are timeouts due to running slow rather than hanging.

DanTup committed:
Don't try to stop the same app twice in flutter_attach test (#20905)

This test is sometimes flaky (#20822) because the two processes try to stop the same app. This fix changes to just gracefully terminate the attach process without explicitly trying to stop and then uses the original spawning process to stop the app.

I can't repro the flake locally to be certain, but I've verified only one stop command is sent now so it *should* be good.

liyuqian committed:
Add customBorder to InkWell so it can clip ShapeBorder (#20751)

This fixes #20483 by letting InkWell do its own clipping.

PathOp.intersect is not used because we have too many unit tests that rely on clipping (e.g., paints..clipXXX()..drawCircle())

The goldens are updated due to small AA changes of the additional clipPath.

amirh committed:
Add support for GestureArenaTeam captains. (#20883)

The team captain wins the arena on behalf of the team.
When any of the team members claims victory for the arena the captain
accepts the gesture.

This is used when embeddeding platform views - we allow configuring a
set of gestures that should be forwarded to the platform view.
We add the set of gesture recognizers to a GestureArenaTeam with a
captain, and if the captain accepts the gesture we forward it to the
platform view.

sir-boformer committed:
Fix small TabBar inconsistencies (#20887)

Tabs are now padded by 16.0 on the left and right  (used to be 12.0).

willlarche committed:
[Gallery] Material elevations (shadows) demo (#20842)

* [Gallery] Elevations demo. TODO: Correct icon.

* [Gallery] More elevations for the elevation demo.

* [Gallery] Elevation toggle for elevation demo.

* [Demos] Update temp icon on elevations demo.

* [Demos] Changing action icon.

* [Demos] Same.

* [Demos] Adding new and const.

* [Demos] Reverting formatting.

* [Gallery] PR Feedback.

Commit:
Add localization for expand icon collapse/expand tap hint (#20766)

a-siva committed:
Roll engine to version 4fde3958388cab4210d6f7da86ec78e32fc4e7ee

Contains the following changes:
4fde39583 Roll Dart to version ccb16f72824374163562364bf19dd18e8a882fab (#6065)
140a5b757 Alllow access to Flutter engine, Dart & Skia versions in the Shell API. (#6060)

a-siva committed:
Roll engine to version c62248bdd3f46794b78d1bd141e92691d539e56f

This includes the following changes
c62248bdd Roll src/third_party/skia 82cf64a0d317..33d5394d08f6 (14 commits) (#6068)
b0fb3e660 Roll src/third_party/skia 96597c22c344..82cf64a0d317 (12 commits) (#6067)
ea8b949a7 Roll src/third_party/skia 4d727c560b58..96597c22c344 (27 commits) (#6066)
4fde39583 Roll Dart to version ccb16f72824374163562364bf19dd18e8a882fab (#6065)
140a5b757 Alllow access to Flutter engine, Dart & Skia versions in the Shell API. (#6060)

Commit:
Mark memory_nav as flaky (#20930)

amirh committed:
Add a layoutDirection parameter to AndroidView (#20838)

engine-flutter-autoroll committed:
Roll engine c62248bdd3f4..0b728a314001 (3 commits) (#20942)

Auto-roller completed checks. Merging.

Commit:
Create flutter_driver infra for testing the Android AccessibilityNodeInfo generated by Flutter (#19700)

zoechi committed:
Fix image refs to moved image directory

fixes #20940

doxuanviet1996 committed:
add cupertino localizations (#20756)

add cupertino localizations abstract class and the default cupertino localizations in US English.

doxuanviet1996 committed:
merges with localizations

doxuanviet1996 committed:
resolves confligts
